app/src/main/java/com/github/geohunt/app/ComposeActivity.kt

```diff
@@ -0,0 +1,121 @@
+
+    /**
+     * Called when the activity is starting
+     */
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_compose)
+
+        val mapFragment =
+                supportFragmentManager.findFragmentById(R.id.map_container_view) as? SupportMapFragment
+        mapFragment?.getMapAsync(this)
+        mapFragment?.getMapAsync{ map ->
+            map.setInfoWindowAdapter(EventMarkerActionDisplay(this))
+        }
+    }
+
+    /**
+     * Called when the map is ready to be used
+     */
+    override fun onMapReady(googleMap: GoogleMap) {
+        map = googleMap
+        addMarkersOnTheMap(googleMap, mockChallengeDatabase)
+        enableLocation()
+
+        val epflLatitude = 46.519585
+        val epflLongitude = 6.5684919
+        val epflCoordinates = LatLng(epflLatitude, epflLongitude)
+
+        val mapZoomValue = 15f
+        map.moveCamera(CameraUpdateFactory.newLatLngZoom(epflCoordinates, mapZoomValue))
+    }
+
+    /**
+     * Adds the data from the database to the map as markers
+     */
+    private fun addMarkersOnTheMap(map: GoogleMap, challenges: List<Triple<Bitmap, LatLng, LocalDateTime>>){
+        challenges.forEach{challenge ->
+                val marker = map.addMarker(
+                MarkerOptions()
+                        .position(challenge.second)
+                )
+            marker?.tag = challenge
+        }
+    }
+
+    /**
+     * Enables the location layer if the permission has been granted
+     * Otherwise, requests the permission
+     */
+    override fun onRequestPermissionsResult(
+            requestCode: Int,
+            permissions: Array<out String>,
+            grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == REQUEST_CODE_LOCATION_PERMISSION && grantResults.isNotEmpty()) {
+            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                enableLocation()
+            }
+        }
+    }
+
+    /**
+     * Checks if the user has granted the location permission
+     * @return true if the permission is granted, false otherwise
+     */
+    private fun isPermissionGranted(): Boolean {
+        return ContextCompat.checkSelfPermission(
+                this,
+                Manifest.permission.ACCESS_FINE_LOCATION
+        ) == PackageManager.PERMISSION_GRANTED
+    }
+
+    /**
+     * Enables the location layer if the permission has been granted
+     * Otherwise, requests the permission
+     */
+    private fun enableLocation() {
+        if (isPermissionGranted()) {
+            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+                return
+            }
```

Comment:
I don't understand the purpose of those 3 lines, since the permission was checked the line above.

Reply:
This is because the call requires permission which may be rejected by user: code should explicitly check to see if permission is available (with checkPermission) or explicitly handle a potential SecurityException. By doing this part of code, I am handling the possible lack of assignment of permissions.

Reply:
I don't understand why the permission checks should be done twice. My question is whether there is a difference between using `ContextCompat` (as done in `isPermissionGranted`) and `ActivityCompat` to check the permissions, and if there is, maybe we should explicitly mention it as a comment in the code ?